python/tests/test_tables.py

```diff
@@ -851,6 +851,61 @@ def test_round_trip_set_columns(self):
                 assert table[j].metadata == metadata_column[j]
                 assert table[j + num_rows].metadata == metadata_column[j]
 
+    def test_set_null_metadata(self):
+        for codec in ["struct", "json"]:
+            table = self.table_class()
+            table.metadata_schema = metadata.MetadataSchema(
+                {
+                    "codec": f"{codec}",
+                    "title": "Example Metadata",
+                    "type": ["object", "null"],
+                    "properties": {
+                        "a": {"type": "number", "binaryFormat": "i"},
+                    },
+                    "required": ["a"],
+                    "additionalProperties": False,
+                },
+            )
+            metadata_list = [{"a": 4}, None, {"a": 5}]
+            for md in metadata_list:
+                table.add_row(
+                    **{
+                        **self.input_data_for_add_row(),
+                        "metadata": md,
+                    }
+                )
+            assert table.num_rows == len(metadata_list)
+            for md, row in zip(metadata_list, table):
+                assert md == row.metadata
+
+    @pytest.mark.skip("Fails, but shouldn't?")
+    def test_set_with_optional_properties(self):
```

Comment:
> Might as well add {} here too to check it doesn't become None with future fiddling.

Good suggestion, @benjeffery! This fails for totally different reasons on both codecs.

Reply:
With `struct` it fails with:
```
tskit/metadata.py:452: in <lambda>
    else b"".join(
>           sub_encoder(obj[key]) for key, sub_encoder in sub_encoders.items()
        )
    )
E   KeyError: 'a'
```
and with `json` it fails by turning the None into `{}`:
```
>               assert md == row.metadata
E               assert None == {}
E                +  where {} = IndividualTableRow(flags=3, location=array([], dtype=float64), parents=array([], dtype=int32), metadata={}).metadata
```